### ISSUE TITLE
Add ability to bypass database creation

### ISSUE THREAD

**Issue by twerthi** [erikbra/grate#211]
**Is your feature request related to a problem? Please describe.**
The wording of the `--adminconnectionstring` is confusing as it states `...if you want to create the database`.  This makes me think that if I set this configuration option to blank, it will bypass the checking and create if not exists portion of the code.  However, setting it to blank fails as it can't find a host to connect to.

I also ran into an issue where, when using MySQL and the user account was not a superuser, the database check would fail with `Uknown database: master` which I thought was odd since I don't believe MySQL has a `master` database and that the database `mysql` fulfilled that.  In addition, the user account that I was testing (Azure Managed Identity) didn't have permissions to query the databases table to determine if the database it was configured to execute against existed so it would return false, then try to create it and fail since the database already existed.

Another issue I ran into was that though the `--adminconnectionstring` states it defaults to the value of `--connectionstring`, it would fail with the above `master` database error.  However, if I explicitly set the same value for `--adminconnectionstring` it would succeed (if the database didn't previously exist).

**Describe the solution you'd like**
The description of `--adminconnectionstring` makes it sound like setting it to blank would bypass the database exists/create functionality.   Alternatively, RoundhousE had a `--donotcreatedatabase` option you could set, implementing that would work as well.

**Describe alternatives you've considered**

**Additional context**

**Comment by wokket** [comment]
Hey @twerthi,

Sorry for the delay getting back to this, my world is crazy right now.

We do [have code support](https://github.com/erikbra/grate/blob/70decb440299ddf3f31ed5bf50b3d81edcc033a5/grate/Configuration/GrateConfiguration.cs#L54) for not creating the database, but it doesn't look be exposed on the commandline yet, oops!

I agree that the wording around `AdminConnectionString` doesn't convey that it's used for all server0level activities (as opoosed to database level), ie both the create and any `AlterDatabase`  scripts that need to be run.  Would you be open to drafting a PR that cleans up the docs in a way that makes more sense to you?

I don't use MySql myself, but the [MariaDb unit tests](https://github.com/erikbra/grate/runs/7248404244#r0s0) include a create test (see  `Is_created_if_confed_and_it_does_not_exist`) that is currently passing 🤷?   There's quite a few moving parts in that issue (incl azure, varying permission levels etc) but that and the final issue may be related to [this code](https://github.com/erikbra/grate/blob/70decb440299ddf3f31ed5bf50b3d81edcc033a5/grate/Configuration/GrateConfiguration.cs#L34-L51).  Again I haven't used MySql in more than 20 years so it's going to take some digging.

**Comment by wokket** [comment]
> When omitting the --adminconnectionstring argument, it would fail with the message about master database not existing. If I set the value of --adminconnectionstring to the same value as --connectionstring it would succeed.

Yep, that aligns with the code I linked above assuming 'master' if the admin string isn't set... At least you have a workaround for now.

I doubt I'll have a chance to look into this much deeper this week, but will try and get a release published for you when I can.

**Comment by twerthi** [comment]
Understood, thank you for letting me know :)  I was wondering if that was the case, I'd perused the code and saw that 'master' part, but wasn't entirely sure I was looking in the right spot.

No worries on getting the release out, I have other things I can work on :)  Not sure if I mentioned it, but testing with a GCP Service Account worked well ;)

**Comment by erikbra** [comment]
Hi, guys. Happy to see that various corner cases are being explored here. I _did_ think exposing the "create/don't create database if it doesn't exist" option wasn't necessary anymore in grate, as we use a different strategy than RoundhousE did on checking whether the database exists or not (try to connect, if it fails with a specific exception, assume it doesn't exists). 

But I might have been wrong.

I originally created the `--donotcreatedatabase` option for RoundhousE, exactly for this use case (with SQL Azure though, but same situation, with no 'master' database access), but I thought it wasn't needed when I implemented this differently in grate.

Do you have a bit more details on the setup you are using with MySql, @twerthi ? I would like to explore this a bit further, to see if the `--createdatabase` _is_ actually needed, or if it was just a bug in the implementation, which didn't work for MySql.

**Comment by twerthi** [comment]
I was using MySQL on Azure testing with an Azure Managed Identity.  The account isn't a superuser so it is unable to query the `mysql` database so the check to see if the database exists always returns false.  The process I was using first created the database  using the admin username/password then gave full rights to the database to the Managed Identity.  This would always fail because it would attempt to create the database again.

**Comment by twerthi** [comment]
Any update on this?  Still blocked on creating the Step Template for Octopus Deploy until I can get around this issue.

**Comment by erikbra** [comment]
Hi, @twerthi , very sorry about the delay here. I have work in the daytime too ;)

I simply cannot get this to reproduce. If I create a user first with MySql (locally, in docker), with rights only on the newly created database, I still get a list of databases, but it only contains the newly created database, and INFORMATION_SCHEMA. I create the following user, with the following grants:

```
CREATE DATABASE {DatabaseName}
CREATE USER '{UserName}'@'%' IDENTIFIED BY '{Password}';
GRANT SELECT, INSERT, UPDATE, DELETE, CREATE, INDEX, DROP, ALTER, CREATE TEMPORARY TABLES, LOCK TABLES ON {DatabaseName}.* TO '{UserName}'@'%';
FLUSH PRIVILEGES;
```


And it just works(tm).

What steps are you performing to create the user, and are you granting the correct privileges to the user before trying to run grate?

**Comment by twerthi** [comment]
First, thank you for taking the time to try to reproduce, wanted to make sure I appreciate it.

In my case, I was using an Azure MySQL server instance along with an Azure Managed Identity account assigned to a VM.  Using the directions from https://docs.microsoft.com/en-us/azure/mysql/single-server/how-to-connect-with-managed-identity, I would add the user account so the managed identity had access to the server.  Next, I would issue ```GRANT CREATE ON *.* TO 'linuxconfig'@'localhost';``` so the user account could create databases and be the owner of them.  The account was able to successfully create the database, however, when grate ran, it kept thinking it wasn't there so it attempted to create it again which failed.  If I didn't create the database prior to running grate, it would succeed.  I'll try and dig up what else I'd tried, but that was the gist of what's going on.

**Comment by erikbra** [comment]
Hi again, @twerthi - of course I need to try to reproduce, I don't want bugs ;)  - thanks for taking the time to report details.

I'm not sure I understand fully (I might be a bit slow), but are you saying that it _does not_ work if you create the database prior to running grate? So it _does work_ if you let grate create the database, using the user you have created?

If I understood it correctly, it does make sense in my head, because if you use the user you have created to create the database, it would probably get rights on the database, but if you create it beforehand, it wouldn't (?)

I'm no MySql expert, by all means, and I might be misunderstanding what you are saying. Which user identity are you using when you create the database prior to running grate?

Even though my theory is that this _should_ work with the current grate, I have prepared a PR which simplifies the way we check if a database exists or not: #227 

Are you comfortable building from scratch, so you could try out the PR version, and see if it fixes your issue or not? If not, I can build a binary for you
